fluentd/event-exporter/sinks/stackdriver/sink.go

```diff
@@ -0,0 +1,163 @@
+			Subsystem: "stackdriver_sink",
+		},
+		[]string{"component", "host"},
+	)
+
+	successfullySentEntryCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name:      "successfully_sent_entry_count",
+			Help:      "Number of entries, successfully ingested by Stackdriver",
+			Subsystem: "stackdriver_sink",
+		},
+	)
+)
+
+type sdSink struct {
+	logEntryChannel chan *sd.LogEntry
+	config          *sdSinkConfig
+	logEntryFactory *sdLogEntryFactory
+	writer          sdWriter
+	logName         string
+
+	currentBuffer      []*sd.LogEntry
+	timer              *time.Timer
+	fakeTimeChannel    chan time.Time
+	concurrencyChannel chan struct{}
+}
+
+func newSdSink(writer sdWriter, clock clock.Clock, config *sdSinkConfig) *sdSink {
+	return &sdSink{
+		logEntryChannel: make(chan *sd.LogEntry, config.MaxBufferSize),
+		config:          config,
+		logEntryFactory: newSdLogEntryFactory(clock),
+		writer:          writer,
+		logName:         config.LogName,
+
+		currentBuffer:      []*sd.LogEntry{},
+		timer:              nil,
+		fakeTimeChannel:    make(chan time.Time),
+		concurrencyChannel: make(chan struct{}, config.MaxConcurrency),
+	}
+}
+
+func (s *sdSink) Add(event *api_v1.Event) {
+	receivedEntryCount.WithLabelValues(event.Source.Component, event.Source.Host).Inc()
+
+	logEntry := s.logEntryFactory.FromEvent(event)
+	s.logEntryChannel <- logEntry
+}
+
+func (s *sdSink) Update(oldEvent *api_v1.Event, newEvent *api_v1.Event) {
+	if newEvent.Count != oldEvent.Count+1 {
+		// Sink doesn't send a LogEntry to Stackdriver, b/c event compression might
+		// indicate that part of the watch history was lost, which may result in
+		// multiple events being compressed. This may create an unecessary
+		// flood in Stackdriver.
+		glog.V(2).Infof("Event count has increased by something but one.\n"+
+			"\tOld event: %s\n\tNew event: %s", oldEvent, newEvent)
+	}
+
+	receivedEntryCount.WithLabelValues(newEvent.Source.Component, newEvent.Source.Host).Inc()
+
+	logEntry := s.logEntryFactory.FromEvent(newEvent)
+	s.logEntryChannel <- logEntry
+}
+
+func (s *sdSink) Delete(*api_v1.Event) {
+	// Nothing to do here
+}
+
+func (s *sdSink) FilterList(events []api_v1.Event) []api_v1.Event {
+	if len(events) > 0 {
+		glog.V(2).Infof("List request returned non-empty response")
+		s.logEntryChannel <- s.logEntryFactory.FromMessage("Some events may have been lost")
+	}
+	return []api_v1.Event{}
+}
+
+func (s *sdSink) Start() {
+	go func() {
+		for {
+			select {
+			case entry := <-s.logEntryChannel:
+				s.currentBuffer = append(s.currentBuffer, entry)
+				if len(s.currentBuffer) >= s.config.MaxBufferSize {
+					s.flushBuffer()
+				} else if len(s.currentBuffer) == 1 {
+					s.setTimer()
+				}
+				break
+			case <-s.getTimerChannel():
+				s.flushBuffer()
+				break
+			}
+		}
+	}()
+}
+
+func (s *sdSink) flushBuffer() {
+	entries := s.currentBuffer
+	s.currentBuffer = nil
+	s.concurrencyChannel <- struct{}{}
```

Comment:
No, that's not true, we want to be able to keep up with the stream of events and one thread may not be enough